### PR TITLE
RFC and Proof of Concept for GitHub GitOps

### PR DESCRIPTION
**Pull request by karlmdavis**
**JIRA Ticket:**
N/A (see the epic and story text files included in this PR)

**User Story or Bug Summary:**

Here's the overall epic for the larger effort:

**For** DASG federal and contractor staff
  **who** need elevated permissions to DASG's GitHub repositories
  **the** `git-ops-for-github` project
  **is a** Terraform project that gets applied via GitHub Actions
  **that** is simple to manage and easy to audit.
**Unlike** the current "find and ask an admin" option,
  **our solution** is more self-service, faster, and more secure.
**Unlike** the other "file a Help Desk ticket" option,
  **our solution** is more self-service, faster, and more secure.

Here's the user story for the proof of concept contained here:

As a DASG techie, I need a workflow and tooling for declaratively managing the permissions of the <https://github.com/CMSgov/cms-oeda-dasg> repository, so that we can turn around requests for GitHub access more quickly and safely.

### What Does This PR Do?

This pull request contains an RFC, design, and proof of concept for managing DASG's GitHub repositories via Terraform, deployed via GitHub Actions.

As a proof-of-concept, the implementation still has a number of intentional gaps, including:

1. User/staff lists are hilariously incomplete.
2. If actually applied, would overwrite a number of repository settings, e.g. branch protection.
3. I haven't actually _tried_ to apply it yet; I've only gone as far as `terraform plan`.
4. Doesn't yet use a remote state backend for Terraform.
5. Teams are created with private membership, which is silly.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check these things, in particular:

* Please review the RFC. Approval on this PR is a vote to approve the RFC.
    * Here's the rendered-as-HTML version of the RFC: [rfcs/0000-git-ops-for-github.md](https://github.com/CMSgov/cms-oeda-dasg/blob/e37d6977a77c029d2f0ff5e605848869a836cabc/rfcs/0000-git-ops-for-github.md).
* Please review the epic and stories for the first post-proof-of-concept implementation goals for this project. Are they sufficient for a first "release"?
* Please review the Terraform code in the proof of concept. It's intentionally incomplete/just-a-proof-of-concept, but I currently expect to be able to use a similar approach, going forwards.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment: N/A.
    * Does this PR add any new software dependencies? **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **Yes**: @StewGoin do you have any concerns with the proposed approach? The RFC and proof of concept themselves don't need a SIA, but at least one of the implementation steps definitely will.

### What Needs to Be Merged and Deployed Before this PR?

N/A

### Submitter Checklist

Before requesting final review, I've gone through and verified that this PR complies with the following requirements:

* [X] I've refactored and rebased this PR (for help, see: [this](https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had) and [this](https://raphaelfabeni.com/git-editing-commits-part-1/), if needed, so that it is as small as it can reasonably be, in order to:
    1. Ensure that any problems it causes have a small "blast radius".
    2. Ensure that it'll be easier to rollback if that becomes necessary.
    3. Ease the burden on reviewers.
* [X] I've verified that the commits in this PR:
    1. Reasonably explain the "what" and "why" of the changes.
    2. Leverage JIRA's [smart commits](https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html) to reference the JIRA ticket that they're associated with. For example, aim for commit messages like this (note also [the 50/72 formatting](https://stackoverflow.com/q/2290016) used here):
        
        ```
        Added the whizbang to the doodad.

        The new whizbang should really simplify things for our whoosit users.
        It was a bit tricky to get it into the doodad, but we decided that the
        flipwhee pattern was the best approach for now. Might want to revisit
        that in the future if it ends up being too hard to maintain.

        SOMEPROJECT-42
        ```
        
* [X] I've verified that this PR includes all required documentation changes, including `README` updates and changelog / release notes entries.
* [X] I've verified that all new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] I've verified that all tech debt and/or shortcomings introduced by this PR is detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] I've requested review from at least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
* [X] I've requested review from any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [X] I've verified that this PR and its changes comply with all other policies in the [DASG Engineering Standards](../policies/engineering_standards.md) and have specifically called out any/all deviations in this PR.

### CODE REVIEW

**Comment by karlmdavis** [comment]
Here's the output of `terraform plan -no-color -var github_token=...` (note that I haven't yet imported any existing state -- need to do that):

```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.cms_program_oeda_dasg.github_repository.repos["cms-oeda-dasg"] will be created
  + resource "github_repository" "repos" {
      + allow_merge_commit     = true
      + allow_rebase_merge     = true
      + allow_squash_merge     = true
      + archived               = false
      + default_branch         = (known after apply)
      + delete_branch_on_merge = false
      + etag                   = (known after apply)
      + full_name              = (known after apply)
      + git_clone_url          = (known after apply)
      + html_url               = (known after apply)
      + http_clone_url         = (known after apply)
      + id                     = (known after apply)
      + name                   = "cms-oeda-dasg"
      + node_id                = (known after apply)
      + ssh_clone_url          = (known after apply)
      + svn_url                = (known after apply)
    }

  # module.cms_program_oeda_dasg.github_team.admins will be created
  + resource "github_team" "admins" {
      + etag    = (known after apply)
      + id      = (known after apply)
      + name    = "oeda-dasg-admins"
      + node_id = (known after apply)
      + privacy = "secret"
      + slug    = (known after apply)
    }

  # module.cms_program_oeda_dasg.github_team.partners will be created
  + resource "github_team" "partners" {
      + etag    = (known after apply)
      + id      = (known after apply)
      + name    = "oeda-dasg-partners"
      + node_id = (known after apply)
      + privacy = "secret"
      + slug    = (known after apply)
    }

  # module.cms_program_oeda_dasg.github_team.staff will be created
  + resource "github_team" "staff" {
      + etag    = (known after apply)
      + id      = (known after apply)
      + name    = "oeda-dasg-staff"
      + node_id = (known after apply)
      + privacy = "secret"
      + slug    = (known after apply)
    }

  # module.cms_program_oeda_dasg.github_team_membership.admins_maintainers["karlmdavis"] will be created
  + resource "github_team_membership" "admins_maintainers" {
      + etag     = (known after apply)
      + id       = (known after apply)
      + role     = "maintainer"
      + team_id  = (known after apply)
      + username = "karlmdavis"
    }

  # module.cms_program_oeda_dasg.github_team_membership.partners_maintainers["karlmdavis"] will be created
  + resource "github_team_membership" "partners_maintainers" {
      + etag     = (known after apply)
      + id       = (known after apply)
      + role     = "maintainer"
      + team_id  = (known after apply)
      + username = "karlmdavis"
    }

  # module.cms_program_oeda_dasg.github_team_membership.staff_maintainers["karlmdavis"] will be created
  + resource "github_team_membership" "staff_maintainers" {
      + etag     = (known after apply)
      + id       = (known after apply)
      + role     = "maintainer"
      + team_id  = (known after apply)
      + username = "karlmdavis"
    }

  # module.cms_program_oeda_dasg.github_team_membership.staff_members["RickHawesUSDS"] will be created
  + resource "github_team_membership" "staff_members" {
      + etag     = (known after apply)
      + id       = (known after apply)
      + role     = "member"
      + team_id  = (known after apply)
      + username = "RickHawesUSDS"
    }

  # module.cms_program_oeda_dasg.github_team_repository.repos_admins["cms-oeda-dasg"] will be created
  + resource "github_team_repository" "repos_admins" {
      + etag       = (known after apply)
      + id         = (known after apply)
      + permission = "admin"
      + repository = "cms-oeda-dasg"
      + team_id    = (known after apply)
    }

  # module.cms_program_oeda_dasg.github_team_repository.repos_partners["cms-oeda-dasg"] will be created
  + resource "github_team_repository" "repos_partners" {
      + etag       = (known after apply)
      + id         = (known after apply)
      + permission = "push"
      + repository = "cms-oeda-dasg"
      + team_id    = (known after apply)
    }

  # module.cms_program_oeda_dasg.github_team_repository.repos_staff["cms-oeda-dasg"] will be created
  + resource "github_team_repository" "repos_staff" {
      + etag       = (known after apply)
      + id         = (known after apply)
      + permission = "maintain"
      + repository = "cms-oeda-dasg"
      + team_id    = (known after apply)
    }

Plan: 11 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```

File: ops/git-ops-for-github/main.tf [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+# The OEDA/DASG (govvie) teams and repos.
+module "cms_program_oeda_dasg" {
+  source       = "./modules/cms_program"
+
+  team_prefix  = "oeda-dasg"
+  repositories = [
+    "cms-oeda-dasg"
+  ]
+
+  admins       = [
```
Comment (msnook) [inline ops/git-ops-for-github/main.tf:20]: Should the elements used to populate `admins`, `staff`, and `partners` be variables?

File: ops/git-ops-for-github/dev/design/0001-story-directory-layout.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Design Record: How to Structure the Directory Layout for this Project?
+
+This is my first TF greenfield project, so I didn't have a lot of experience with this.
+However, I've heard enough chatter and read enough articles about it to know that it's a matter of some debate.
+TF's docs have a page detailing some of the common options:
+  [Terraform Docs: Repository Structure](https://www.terraform.io/docs/cloud/workspaces/repo-structure.html).
+
+Apparently, Terraform provides a "workspaces" mechanism for managing multiple environments.
```
Comment (msnook) [inline ops/git-ops-for-github/dev/design/0001-story-directory-layout.md:8]: I agree - stay away from "workspaces".

File: ops/git-ops-for-github/main.tf (hunk 1 above)
Comment (karlmdavis) [reply ops/git-ops-for-github/main.tf:20]: Mebbe eventually. Don't have a use case for it, yet, so I'm inclined to leave it local-and-easier-to-read unless/until I need to reuse them, or supply them at runtime, or whatever.

Any reason you can see, in advance of that?

**Review by StewGoin — APPROVED** [review]
Looks, to me in my rudimentary terraform knowledge, reasonable as a PoC.

I would request one limitation: prior to genuinely depending on this for operational needs/etc.. let's make sure we get remote state and such so we're not fully dependent on both one person and one physical device where the state file lives.

**Comment by karlmdavis** [comment]
> I would request one limitation: prior to genuinely depending on this for operational needs/etc.. let's make sure we get remote state and such so we're not fully dependent on both one person and one physical device where the state file lives.

Strong agree! That requirement is captured here: https://github.com/CMSgov/cms-oeda-dasg/pull/3/files#diff-818c5944a0e5d2a39aa340d280a79896

I do not plan to `terraform apply` anything until that's in place. In fact, I don't plan to `terraform apply` locally, ever -- I want that handled by a GitHub Action (as detailed in https://github.com/CMSgov/cms-oeda-dasg/pull/3/files#diff-e8346309c263e9dcb1c7771a2b05824b).

**Comment by karlmdavis** [comment]
@StewGoin, @albertico-gov, and @SJWalter11: I've addressed the feedback from our discussion. Can you please review ASAP?

**Comment by StewGoin** [comment]
> @StewGoin, @albertico-gov, and @SJWalter11: I've addressed the feedback from our discussion. Can you please review ASAP?

Still approved from my perspective 👍

**Comment by karlmdavis** [comment]
Just calling the play, for transparency: I’m going to merge this PR now, even without two approved reviews. I’ve addressed the actionable feedback to the best of my ability and I know folks are busy what with [gestures at the world, in general] so I’m going to let things move forward, cautiously.

Thinking back to our discussion on the phone some weeks back, I know @albertico-gov had concerns around usability for non-engineers and also around repo configurability. I’ll try to keep an eye out for signals that those are causing pain (frequent Terraform conflicts or frequent repo config PRs would indicate this) and adjust course, as necessary. I’m not at all wedded to this solution, just committed to finding a good solution.

**Comment by SJWalter11** [comment]
This looks good to me as a start to move to user automation for GitHub.   With the assumed expectation that technical users are working with it.

**Pull request merged**